src/compat/posix/fnmatch/fnmatch.c

```diff
@@ -0,0 +1,160 @@
+/**
+ * @file
+ * @brief fnmatch_function
+ *
+ * @date 20.12.18
+ * @author Filipp Chubukov
+ */
+
+#include <stdlib.h>
+#include <errno.h>
+#include <fnmatch.h>
+
+static int bracket_check(const char *pattern) {
+	while (*pattern != ']') {
+		if (*pattern == '/') {
+			return 0;
+		}
+		pattern++;
+	}
+	return 1;
+}
+
+static int fnmatch_string(const char *pattern, const char *string, int flags) {
+	int flag = 0;
+	int denial_flag = 0;
+	while (*pattern != '\0') {
+		flag = 0;
+		denial_flag = 0;
+
+		if ((*string == '\0') && (*pattern != '*')) {
+			return FNM_NOMATCH;
+		}
+		switch (*pattern) {
+
+			case '?':
+				break;  
+
+			case '*':
+				pattern++;
+				if (*pattern == '\0') {
+					return 0;
+				}
+				while (*string != '\0') {
+					if (!fnmatch_string(pattern, string, flags)) {
+						return 0;
+					}
+					string++;
+				}
+				return FNM_NOMATCH;
+
+			case '/':
+				if (*pattern != *string) {
+					return FNM_NOMATCH;
+				}
+				if (flags == FNM_PATHNAME) {
+					return 0;
+				}
+				break;		
+
+			case '[':
+				if (flags == FNM_PATHNAME) {
+					if (!bracket_check(pattern)) {
+						if (*pattern != *string) {
+							return FNM_NOMATCH;
+						}
+						pattern++;
+						string++;
+						return fnmatch_string(pattern, string, flags);
+					}
+				}	
+
+				pattern++;
+				if (*pattern == '^') {
+					denial_flag = 1;
+					pattern++;
+				}
+				while (*pattern != ']') {
+					if (*string == *pattern) {
+						flag = 1;
+					}
+					pattern++;
+				}
+
+				if ((!flag && !denial_flag) || (flag && denial_flag)) {
+					return FNM_NOMATCH;
+				}
+				break;
+
+			default:
+				if (*pattern != *string) {
+					return FNM_NOMATCH;
+				}
+
+		}
+		string++;
+		pattern++;
+	}
+	if (*string != '\0') {
+		return FNM_NOMATCH;
+	}
+	return 0;
+}
+
+int fnmatch(const char *pattern, const char *string, int flags) {
+	int bracket_open = 0;
+	const char *pattern_check = pattern;
+
+	if (pattern == NULL || string == NULL) {
+		return EINVAL;
+	}
+	while (*pattern_check != '\0') { /* Beginning of pattern-check */
+		if (*pattern_check == '[') {
+			bracket_open = 1;
+		}
+
+		if (*pattern_check == ']') {
+			bracket_open = 0;
+		}
+		pattern_check++;
+	}
+
+	if (bracket_open == 1) {
+		return EINVAL;
+	}				 /* End of pattern-check */
+
+	/* fnmatch for strings */
+	if ((!(flags & FNM_PATHNAME))) {		
+		return fnmatch_string(pattern, string, 0);
+	}
+
+	/* fnmatch for filepaths */
+	if ((*pattern == '/') && (*string == '/')) {
```

Comment:
Do you really need this check? It seems that almost the same thing is done in `while ()`-loop few lines below.

Reply:
its need for first check